src/Data/Singletons/Promote/Defun.hs

```diff
@@ -340,279 +386,496 @@ defunctionalize name m_fixity m_arg_tvbs' m_res_kind' = do
+What should the kind of QuuxSym0 be? Intuitively, it should be this:
 
-         { k :-> DVarT k                                         -- k
-         , t :-> DSigT (DVarT t) (DVarT k)                       -- (t :: k)
-         , a :-> DSigT (DVarT a) (DConT ''Proxy `DAppT` DVarT t) -- (a :: Proxy t)
-         , y :-> DSigT (DVarT y) (DConT ''Proxy `DAppT` DVarT y) -- (y :: Proxy t)
-         , e :-> DSigT (DVarT e) (DConT ''(:~:)
-                                  `DAppT` DSigT (DVarT a) (DConT ''Proxy `DAppT` DSigT (DVarT t) (DVarT k))
-                                  `DAppT` DSigT (DVarT y) (DConT ''Proxy `DAppT` DSigT (DVarT t) (DVarT k)))
-                                                                 -- (e :: (a :: Proxy (t :: k)) :~: (y :: Proxy (t :: k)))
-         }
+  type QuuxSym0 :: forall (k :: Type) ~> k ~> Type
 
-       Why do this? Because when constructing the `forall` in the return kind
-       of a defunctionalization symbol, it's convenient to be able to know
-       the kinds of everything being bound at a glance. It's not always
-       possible to recover the kinds of every variable (for instance, if
-       we're just given `Proxy t ~> Proxy t ~> (a :~: y) ~> Type`), so having
-       this information is handy.
+Alas, `forall (k :: Type) ~>` simply doesn't work. See #304. But there is an
+acceptable compromise we can make that can give us defunctionalization symbols
+for Quux. Once again, we fall back to the partial kind algorithm:
 
-       To construct this map, we:
+  data QuuxSym0 :: Type ~> k ~> Type where ...
+  data QuuxSym1 (k :: Type) :: k ~> Type where ...
+  type QuuxSym2 (k :: Type) (x :: k) = Quux k x :: Type
 
-       (a) Grab the list of type variable binders (this is given as an input
-           to defunctionalize, as discussed in part (1)) and turn it into a list
-           of types. Also include the return kind (if there is one) in this
-           list, as it may also mention type variables with explicit kinds.
-       (b) Construct a flat list of all type variables mentioned in this list.
-           This may involve looking in the kinds of type variables binders.
-           (Note that this part is crucial—the the Singletons/PolyKinds test
-           will fail to compile without it!)
-       (c) Take the flat list and insert each variable into the map by
-           mapping its name to its type (as demonstrated above).
+The catch is that the kind of QuuxSym0, `forall k. Type ~> k ~> Type`, is
+slightly more general than it ought to be. In practice, however, this is
+unlikely to be a problem as long as you apply QuuxSym0 to arguments of the
+right kinds.
 
-       To continue the Symmetry example:
+Note [No SAKs for fully saturated defunctionalization symbols]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+When generating defunctionalization symbols, most of the symbols are data
+types. The last one, however, is a type family. For example, this code:
```

Comment:
Do you mean that the last one is a type *synonym*?

Reply:
Eep. Yes, I did. Fixed.